<launch>

  <!-- run nodes on turtlebot machine -->
  <arg name="robot" default="-1" doc="sets address for machine tag"/>

  <!-- debug argument -->
  <arg name="debug" default="false" doc="provides your slam node with fake data with known data association"/>

  <!--  option to launch turtlebot in empty world -->
  <arg name="empty_world" default="false" doc="launch turtlebot in empty world"/>



  <!-- turtlebot3 diff drive params -->
  <rosparam command="load" file="$(find nuturtle_description)/config/diff_params.yaml" />


  <!-- basic remote -->
  <include file = "$(find nuturtle_robot)/launch/basic_remote.launch" >
    <arg name="robot" value="$(arg robot)" />
  </include>


  <!-- turtlebot model in world -->
  <group if="$(eval arg('robot') == -1)">
    <include file = "$(find nuturtle_gazebo)/launch/diff_drive_gazebo.launch" >
      <arg name="empty_world" value="$(arg empty_world)" />
    </include>
  </group>


  <!-- teleop -->
  <group if="$(eval arg('robot') != 0)">
    <node pkg="turtlebot3_teleop" type="turtlebot3_teleop_key"
       name="turtlebot3_teleop_keyboard"  output="screen" />
  </group>


  <!-- feature detection -->
  <node machine="turtlebot" name="landmarks" pkg="nuslam" type="landmarks" output="screen" >
    <param name="frame_id" value="base_scan" />
  </node>


  <!-- publish real landmarks and robot pose from gazebo -->
  <group ns="real">
    <node machine="turtlebot" name="analysis" pkg="nuslam" type="analysis" output="screen" >
      <param name="radius" value="0.6" />
      <param name="frame_id" value="base_scan" />
    </node>
  </group>


  <!-- mapping feature markers -->
  <node machine="turtlebot" name="draw_map" pkg="nuslam" type="draw_map" output="screen" >
    <remap if="$(arg debug)" from="landmarks" to="real/landmarks" />
  </node>



  <!-- load params in these are private to the node  -->
  <node machine="turtlebot" name="odometer" pkg="rigid2d" type="odometer" output="screen" >
    <param name="odom_frame_id" value="odom" />
    <param name="body_frame_id" value="base_link" />
    <param name="left_wheel_joint" value="left_wheel_axle" />
    <param name="right_wheel_joint" value="right_wheel_axle" />
  </node>


  <!-- Interface with turtlebot3 -->
  <node machine="turtlebot" name="turtle_interface" pkg="nuturtle_robot" type="turtle_interface" output="screen" >
    <param name="left_wheel_joint" value="left_wheel_axle" />
    <param name="right_wheel_joint" value="right_wheel_axle" />
  </node>


  <!-- rviz -->
  <group if="$(eval arg('robot') != 0)">
  <node name="rviz" pkg="rviz" type="rviz" required="true" args="-d $(find tsim)/config/diff_drive_odom.rviz"/>
  </group>


  <!-- load model into parameter server -->
  <group if="$(eval arg('robot') != -1)">
    <param name="robot_description" command="$(find xacro)/xacro '$(find nuturtle_description)/urdf/diff_drive.urdf.xacro'"/>
  </group>


  <!-- robot state publisher -->
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>


</launch>
